complexity of monitor.ts

### Describe what should be investigated or refactored
```bash
  45:28  warning  Async arrow function has a complexity of 16. Maximum allowed is 10  complexity
```

--- a/src/cli/monitor.test.ts
+++ b/src/cli/monitor.test.ts
@@ -1,9 +1,34 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: 2023-Present The Pepr Authors
 
-import { expect, describe, it, jest } from "@jest/globals";
-import { getLabelsAndErrorMessage, getK8sLogFromKubeConfig } from "./monitor";
+import { expect, describe, it, jest, beforeEach, afterEach } from "@jest/globals";
+import {
+  getLabelsAndErrorMessage,
+  getK8sLogFromKubeConfig,
+  processMutateLog,
+  processValidateLog,
+} from "./monitor";
 import { KubeConfig, Log as K8sLog } from "@kubernetes/client-node";
+import { SpiedFunction } from "jest-mock";
+
+const payload = {
+  level: 30,
+  time: 1733751945893,
+  pid: 1,
+  hostname: "test-host",
+  uid: "test-uid",
+  namespace: "test-namespace",
+  name: "test-name",
+  res: {
+    allowed: true,
+    uid: "test-uid",
+    patch: btoa(JSON.stringify({ key: "value" })),
+    patchType: "test-patch-type",
+    warning: "test-warning",
+    status: { message: "test-message" },
+  },
+  msg: "Check response",
+};
 
 jest.mock("@kubernetes/client-node", () => {
   const mockKubeConfig = jest.fn();
@@ -45,5 +70,98 @@ describe("getLabelsAndErrorMessage", () => {
   ])("should return labels and error message", (uuid, expected) => {
     const result = getLabelsAndErrorMessage(uuid);
     expect(result).toEqual(expected);
+  });
+});
+
+describe("processMutateLog", () => {
+  let consoleLogSpy: SpiedFunction<{
+    (...data: unknown[]): void;
+    (message?: unknown, ...optionalParams: unknown[]): void;
+  }>;
+
+  beforeEach(() => {
+    consoleLogSpy = jest.spyOn(console, "log").mockImplementation(() => {});
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  it("should log a mutation approval with patch details", () => {
+    processMutateLog(
+      {
+        ...payload,
+        res: {
+          ...payload.res,
+          patch: btoa(JSON.stringify({ key: "value" })),
+          patchType: "JSONPatch",
+        },
+      },
+      "test-name",
+      "test-uid",
+    );
+
+    expect(consoleLogSpy).toHaveBeenCalledWith("\n🔀  MUTATE     test-name (test-uid)");
+    expect(consoleLogSpy).toHaveBeenCalledWith(
+      expect.stringContaining(JSON.stringify({ key: "value" }, null, 2)),
+    );
+  });
+
+  it("should log a mutation denial without patch details", () => {
+    processMutateLog(
+      {
+        ...payload,
+        res: {
+          ...payload.res,
+          allowed: false,
+          patch: btoa(JSON.stringify("something")),
+        },
+      },
+      "test-name",
+      "test-uid",
+    );
+
+    expect(consoleLogSpy).toHaveBeenCalledWith("\n🚫  MUTATE     test-name (test-uid)");
+    expect(consoleLogSpy).not.toHaveBeenCalledWith(expect.stringContaining("{"));
+  });
+});
+
+describe("processValidateLog", () => {
+  let consoleLogSpy: SpiedFunction<{
+    (...data: unknown[]): void;
+    (message?: unknown, ...optionalParams: unknown[]): void;
+  }>;
+
+  beforeEach(() => {
+    consoleLogSpy = jest.spyOn(console, "log").mockImplementation(() => {});
+  });
+
+  afterEach(() => {
+    jest.restoreAllMocks();
+  });
+
+  it("should log a successful validation", () => {
+    processValidateLog(payload, "test-name", "test-uid");
+
+    expect(consoleLogSpy).toHaveBeenCalledWith("\n✅  VALIDATE   test-name (test-uid)");
+    expect(consoleLogSpy).not.toHaveBeenCalledWith(expect.stringContaining("❌"));
+  });
+
+  it("should log a validation failure with error messages", () => {
+    processValidateLog(
+      {
+        ...payload,
+        res: {
+          ...payload.res,
+          allowed: false,
+          status: { message: "Failure message 1" },
+        },
+      },
+      "test-name",
+      "test-uid",
+    );
+
+    expect(consoleLogSpy).toHaveBeenCalledWith("\n❌  VALIDATE   test-name (test-uid)");
+    expect(consoleLogSpy).toHaveBeenCalledWith(expect.stringContaining("Failure message 1"));
   });
 });

--- a/src/cli/monitor.test.ts
+++ b/src/cli/monitor.test.ts
@@ -1,0 +1,49 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: 2023-Present The Pepr Authors
+
+import { expect, describe, it, jest } from "@jest/globals";
+import { getLabelsAndErrorMessage, getK8sLogFromKubeConfig } from "./monitor";
+import { KubeConfig, Log as K8sLog } from "@kubernetes/client-node";
+
+jest.mock("@kubernetes/client-node", () => {
+  const mockKubeConfig = jest.fn();
+  mockKubeConfig.prototype.loadFromDefault = jest.fn();
+  const mockK8sLog = jest.fn();
+
+  return {
+    KubeConfig: mockKubeConfig,
+    Log: mockK8sLog,
+  };
+});
+
+describe("getK8sLogFromKubeConfig", () => {
+  it("should create a K8sLog instance from the default KubeConfig", () => {
+    const result = getK8sLogFromKubeConfig();
+    expect(KubeConfig).toHaveBeenCalledTimes(1);
+    expect(KubeConfig.prototype.loadFromDefault).toHaveBeenCalledTimes(1);
+
+    const kubeConfigInstance = new KubeConfig();
+    expect(K8sLog).toHaveBeenCalledWith(kubeConfigInstance);
+
+    expect(result).toBeInstanceOf(K8sLog);
+  });
+});
+
+describe("getLabelsAndErrorMessage", () => {
+  it.each([
+    [
+      undefined,
+      {
+        labels: ["pepr.dev/controller", "admission"],
+        errorMessage: "No pods found with admission labels",
+      },
+    ],
+    ["test", { labels: ["app", "pepr-test"], errorMessage: "No pods found for module test" }],
+    ["test2", { labels: ["app", "pepr-test2"], errorMessage: "No pods found for module test2" }],
+    ["test3", { labels: ["app", "pepr-test3"], errorMessage: "No pods found for module test3" }],
+    ["test4", { labels: ["app", "pepr-test4"], errorMessage: "No pods found for module test4" }],
+  ])("should return labels and error message", (uuid, expected) => {
+    const result = getLabelsAndErrorMessage(uuid);
+    expect(result).toEqual(expected);
+  });
+});

--- a/src/cli/monitor.ts
+++ b/src/cli/monitor.ts
@@ -1,92 +1,51 @@
 // SPDX-License-Identifier: Apache-2.0
 // SPDX-FileCopyrightText: 2023-Present The Pepr Authors
 
-import { Log as K8sLog, KubeConfig } from "@kubernetes/client-node";
+import { Log as K8sLog, KubeConfig, KubernetesListObject } from "@kubernetes/client-node";
 import { K8s, kind } from "kubernetes-fluent-client";
 import stream from "stream";
 import { ResponseItem } from "../lib/types";
 import { RootCmd } from "./root";
 
-export default function (program: RootCmd) {
+interface LogPayload {
+  namespace: string;
+  name: string;
+  res: {
+    uid: string;
+    allowed?: boolean;
+    patch?: string;
+    patchType?: string;
+    warnings?: string;
+    status?: {
+      message: string;
+    };
+  };
+}
+
+export default function (program: RootCmd): void {
   program
     .command("monitor [module-uuid]")
     .description("Monitor a Pepr Module")
     .action(async uuid => {
-      let labels: string[];
-      let errorMessage: string;
-
-      if (!uuid) {
-        labels = ["pepr.dev/controller", "admission"];
-        errorMessage = `No pods found with admission labels`;
-      } else {
-        labels = ["app", `pepr-${uuid}`];
-        errorMessage = `No pods found for module ${uuid}`;
-      }
+      const { labels, errorMessage } = getLabelsAndErrorMessage(uuid);
 
       // Get the logs for the `app=pepr-${module}` or `pepr.dev/controller=admission` pod selector
-      const pods = await K8s(kind.Pod)
+      const pods: KubernetesListObject<kind.Pod> = await K8s(kind.Pod)
         .InNamespace("pepr-system")
         .WithLabel(labels[0], labels[1])
         .Get();
 
-      const podNames = pods.items.flatMap(pod => pod.metadata!.name) as string[];
+      // Pods will ways have a metadata and name fields
+      const podNames: string[] = pods.items.flatMap(pod => pod.metadata!.name || "");
 
       if (podNames.length < 1) {
         console.error(errorMessage);
         process.exit(1);
       }
 
-      const kc = new KubeConfig();
-      kc.loadFromDefault();
+      const log = getK8sLogFromKubeConfig();
 
-      const log = new K8sLog(kc);
-
-      const logStream = new stream.PassThrough();
-      logStream.on("data", async chunk => {
-        const respMsg = `"msg":"Check response"`;
-        // Split the chunk into lines
-        const lines = await chunk.toString().split("\n");
-
-        for (const line of lines) {
-          // Check for `"msg":"Hello Pepr"`
-          if (!line.includes(respMsg)) continue;
-          try {
-            const payload = JSON.parse(line.trim());
-            const isMutate = payload.res.patchType || payload.res.warnings;
-
-            const name = `${payload.namespace}${payload.name}`;
-            const uid = payload.res.uid;
-
-            if (isMutate) {
-              const plainPatch =
-                payload.res?.patch !== undefined && payload.res?.patch !== null
-                  ? atob(payload.res.patch)
-                  : "";
-
-              const patch = plainPatch !== "" && JSON.stringify(JSON.parse(plainPatch), null, 2);
-              const patchType = payload.res.patchType || payload.res.warnings || "";
-              const allowOrDeny = payload.res.allowed ? "🔀" : "🚫";
-              console.log(`\n${allowOrDeny}  MUTATE     ${name} (${uid})`);
-              patchType.length > 0 && console.log(`\n\u001b[1;34m${patch}\u001b[0m`);
-            } else {
-              const failures = Array.isArray(payload.res) ? payload.res : [payload.res];
-
-              const filteredFailures = failures
-                .filter((r: ResponseItem) => !r.allowed)
-                .map((r: ResponseItem) => r.status.message);
-
-              console.log(
-                `\n${filteredFailures.length > 0 ? "❌" : "✅"}  VALIDATE   ${name} (${uid})`,
-              );
-              console.log(
-                filteredFailures.length > 0 ? `\u001b[1;31m${filteredFailures}\u001b[0m` : "",
-              );
-            }
-          } catch {
-            // Do nothing
-          }
-        }
-      });
+      const logStream = createLogStream();
 
       for (const podName of podNames) {
         await log.log("pepr-system", podName, "server", logStream, {
@@ -96,4 +55,88 @@ export default function (program: RootCmd) {
         });
       }
     });
+}
+
+export function getLabelsAndErrorMessage(uuid?: string): {
+  labels: string[];
+  errorMessage: string;
+} {
+  let labels: string[];
+  let errorMessage: string;
+
+  if (!uuid) {
+    labels = ["pepr.dev/controller", "admission"];
+    errorMessage = `No pods found with admission labels`;
+  } else {
+    labels = ["app", `pepr-${uuid}`];
+    errorMessage = `No pods found for module ${uuid}`;
+  }
+
+  return { labels, errorMessage };
+}
+
+export function getK8sLogFromKubeConfig(): K8sLog {
+  const kc = new KubeConfig();
+  kc.loadFromDefault();
+  return new K8sLog(kc);
+}
+
+export function createLogStream(): stream.PassThrough {
+  const logStream = new stream.PassThrough();
+
+  logStream.on("data", async chunk => {
+    const lines = chunk.toString().split("\n");
+    const respMsg = `"msg":"Check response"`;
+
+    for (const line of lines) {
+      if (!line.includes(respMsg)) continue;
+      processLogLine(line);
+    }
+  });
+
+  return logStream;
+}
+
+export function processLogLine(line: string): void {
+  try {
+    const payload: LogPayload = JSON.parse(line.trim());
+    const isMutate = payload.res.patchType || payload.res.warnings;
+    const name = `${payload.namespace}${payload.name}`;
+    const uid = payload.res.uid;
+
+    if (isMutate) {
+      processMutateLog(payload, name, uid);
+    } else {
+      processValidateLog(payload, name, uid);
+    }
+  } catch {
+    // Do nothing
+  }
+}
+
+export function processMutateLog(payload: LogPayload, name: string, uid: string): void {
+  const plainPatch =
+    payload.res.patch !== undefined && payload.res.patch !== null ? atob(payload.res.patch) : "";
+
+  const patch = plainPatch !== "" && JSON.stringify(JSON.parse(plainPatch), null, 2);
+  const patchType = payload.res.patchType || payload.res.warnings || "";
+  const allowOrDeny = payload.res.allowed ? "🔀" : "🚫";
+
+  console.log(`\n${allowOrDeny}  MUTATE     ${name} (${uid})`);
+  if (patchType.length > 0) {
+    console.log(`\n\u001b[1;34m${patch}\u001b[0m`);
+  }
+}
+
+export function processValidateLog(payload: LogPayload, name: string, uid: string): void {
+  const failures = Array.isArray(payload.res) ? payload.res : [payload.res];
+
+  const filteredFailures = failures
+    .filter((r: ResponseItem) => !r.allowed)
+    .map((r: ResponseItem) => r.status?.message || "");
+
+  console.log(`\n${filteredFailures.length > 0 ? "❌" : "✅"}  VALIDATE   ${name} (${uid})`);
+  if (filteredFailures.length > 0) {
+    console.log(`\u001b[1;31m${filteredFailures}\u001b[0m`);
+  }
 }

--- a/src/cli/monitor.ts
+++ b/src/cli/monitor.ts
@@ -81,7 +81,7 @@ export function getK8sLogFromKubeConfig(): K8sLog {
   return new K8sLog(kc);
 }
 
-export function createLogStream(): stream.PassThrough {
+function createLogStream(): stream.PassThrough {
   const logStream = new stream.PassThrough();
 
   logStream.on("data", async chunk => {
@@ -97,7 +97,7 @@ export function createLogStream(): stream.PassThrough {
   return logStream;
 }
 
-export function processLogLine(line: string): void {
+function processLogLine(line: string): void {
   try {
     const payload: LogPayload = JSON.parse(line.trim());
     const isMutate = payload.res.patchType || payload.res.warnings;

